src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs

```diff
@@ -76,7 +88,59 @@ public static string Encode(byte[] inArray, int offset, int length)
             if (inArray == null)
                 throw LogHelper.LogArgumentNullException("inArray");
 
-            return EncodeString(Convert.ToBase64String(inArray, offset, length));
+            int i;
+
+            char[] table = s_base64Table;
+
+
+            int j = 0;
+
+            int lengthmod3 = length % 3;
+            int limit = offset + (length - lengthmod3);
+            char[] output = new char[(length + 2) / 3 * 4];
+
+            for (i = offset; i < limit; i += 3)
+            {
+                byte d0 = inArray[i];
+                byte d1 = inArray[i + 1];
+                byte d2 = inArray[i + 2];
+
+                output[j + 0] = table[(d0 & 0xfc) >> 2];
+                output[j + 1] = table[((d0 & 0x03) << 4) | ((d1 & 0xf0) >> 4)];
+                output[j + 2] = table[((d1 & 0x0f) << 2) | ((d2 & 0xc0) >> 6)];
+                output[j + 3] = table[d2 & 0x3f];
+                j += 4;
+            }
+
+            //Where we left off before
+            i = limit;
+
+            switch (lengthmod3)
```

Comment:
nit: log an error on the default case, or Debug.Fail

Reply:
The variable has 3 potential values (given that its value = length%3). In the case the value being 0, it wouldn't denote an error, but a case where no more operations are needed. Perhaps this would be more descriptive with if conditions?

Reply:
Oops, my bad, just skimmed wrong. switch statement is fine, might help readability a tiny bit to just comment the default case but I think this is fine how you have it.